Move some uihandler methods to another files and add their test.

olimpias opened the pull request:
Two uihandler methods are moved and their tests are added.

Commit:
Uihandler methods are moved and their tests are added.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approvers: 

Assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/fortio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
Hi @olimpias. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

olimpias commented:
Hi @ldemailly . I moved AddHttps to http_utils file and SetHostAndPort method to network file. Also added their test methods. I think there is a bug on SetHostPort. You have asked it in closed PR. The problem, even I add ip address to its input, the result returns with localhost. For example, I input 8080, 192.168.31.50:8080 to SetHostAndPort method. It returns localhost:8080.  Please let me know, which methods should I move to http_utils in the uihandler.go file. Thanks

ldemailly commented on fhttp/http_test.go:
```diff
@@ -941,6 +941,24 @@ func TestChangeMaxPayloadSize(t *testing.T) {
+	for _, test := range tests {
+		output := AddHTTPS(test.input)
+		if output != test.expected {
+			t.Errorf("%s was expected but %s is received ", test.expected, output)
```
go convention is to print actual first, before expected (I don't think it's very good or important but so far it's been consistently applied)

ldemailly commented on fhttp/http_test.go:
```diff
@@ -941,6 +941,24 @@ func TestChangeMaxPayloadSize(t *testing.T) {
+		{"www.google.com", "https://www.google.com"},
+		{"http://www.google.com", "https://www.google.com"},
+		{"https://www.google.com", "https://www.google.com"},
+	}
```
in theory we'd add mixed case e.g `hTTPs://foo.com/` but I don't think it'll pass in that case so not yet

olimpias replied:
To pass it, I added 	pURL = strings.ToLower(pURL) in the beginning of the AddHTTPS method.

ldemailly replied:
that would break things (url params are case sensitive)

olimpias replied:
While I was doing that I didn't think the param cases ;). I have added a method and test cases for it as well.

ldemailly reviewed (commented):
will check hostandport and your question later, but for starter:

Commit:
Input datas and error logs are updated.

Commit:
Uihandler methods are moved and their tests are added.

ldemailly commented on fhttp/http_test.go:
```diff
@@ -941,6 +941,28 @@ func TestChangeMaxPayloadSize(t *testing.T) {
+		{"foo.com", "https://foo.com"},
+		{"http://foo.com", "https://foo.com"},
+		{"https://foo.com", "https://foo.com"},
+		{"hTTps://foo.com", "https://foo.com"},
```
add /BaR there to confirm when it's already https it's also preserving the case
(and/or fOO.com)

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +451,39 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
+
+// toLowerCasePrefix makes prefix lower case in s. If prefix is not found in s
+// returns input s and false otherwise, returns s with lowered case prefix
+func toLowerCasePrefix(s string, prefix string) (string, bool) {
```
this is unnecessarily expensive/complex

use strings.EqualFold() instead

ldemailly commented on fnet/network.go:
```diff
@@ -160,3 +160,20 @@ func Proxy(port string, dest *net.TCPAddr) *net.TCPAddr {
+// Join the input Host And Port in the form of hostname:port, ip:port or :port. And returns
+func JoinHostAndPort(inputPort string, addr *net.TCPAddr) string {
+	urlHostPort := inputPort
+	portStr := ""
```
this used to be
```
portStr := inputPort
```
in the original code, I doubt it works the same without it ?

olimpias replied:
I used it for different purpose. portStr indicates the port that gathered from net.TCPAddr

ldemailly commented on fnet/network_test.go:
```diff
@@ -169,6 +169,37 @@ func TestResolveIpV6(t *testing.T) {
+			IP:   []byte{192, 168, 30, 14},
+			Port: 8080,
+		}, "192.168.30.14:8080"},
+		{":8080",
```
missing the case for say "8080" (no : ) and nil address
(which will fail to do localhost:8080 )

ldemailly replied:
remove the nil case but do keep 8080 :8080 and [::1]:8080 as inputPort tests

olimpias committed:
Add Https method is updated and some test cases are added.

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
 	_ = o.AddAndValidateExtraHeader("X-On-Behalf-Of: " + r.RemoteAddr)
 }
+
+const (
```
we have those constants a few places already, notably fgrpc/grpcrunner.go - I guess we can clean that up later / put them in a common place without adding dependencies (maybe fnet/)

ldemailly commented on fnet/network.go:
```diff
@@ -160,3 +160,20 @@ func Proxy(port string, dest *net.TCPAddr) *net.TCPAddr {
 }
+
+// Join the input Host And Port in the form of hostname:port, ip:port or :port. And returns
+func JoinHostAndPort(inputPort string, addr *net.TCPAddr) string {
```
I checked the caller in uihandler; addr is never nil so the code simplifies to:
```golang
	urlHostPort := addr.String()
	if strings.HasPrefix(inputPort, ":") || !strings.Contains(inputPort, ":") {
            urlHostPort = fmt.Sprintf("localhost:%d", addr.Port)
	}
	return urlHostPort
```

let's rename this function to `NormalizeHostPort` and the comment to
```
// NormalizeHostPort generates host:port string for the address or uses localhost instead of [::] 
// when the original port binding input didn't specify an address
```

ps: the linter should have complained that the comment doesn't match the function

ldemailly commented on fhttp/http_test.go:
```diff
@@ -941,6 +941,31 @@ func TestChangeMaxPayloadSize(t *testing.T) {
+		input    string
+		expected string
+	}{
+		{"foo.com", "https://foo.com"},
```
add a test for `foo` (shorter than prefix) and maybe `http` which should turn into https://http (if a host name was http)

olimpias replied:
I might miss understand your comment here about shorter than prefix :).

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
+// if url does not contain prefix "http://".
+func AddHTTPS(url string) (pURL string) {
+	if len(url) > len(httpText) {
+		if strings.EqualFold(url[0:len(httpText)], httpText) {
```
remove `0` it's the default (afaik)

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
 }
+
+const (
+	httpText  = "http://"
```
do rename to `httpPrefix` and `httpsPrefix` respectively

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
+
+// addHTTPS replaces "http://" in url with "https://" or prepends "https://"
+// if url does not contain prefix "http://".
+func AddHTTPS(url string) (pURL string) {
```
remove pURL just string return, no need to name it specially given it's not always what is returned

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
+		if strings.EqualFold(url[0:len(httpText)], httpText) {
+			log.Infof("Replacing http scheme with https for url: %s", url)
+			pURL = url[len(httpText):]
+			return httpsText + pURL
```
just directly
```
return httpsPrefix + url[len(httpText):]
```

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
+		// return url unchanged since it already has "https://"
+		if strings.EqualFold(url[0:len(httpsText)], httpsText) {
+			pURL = url[len(httpsText):]
+			return httpsText + pURL
```
same and update the comment that we return normalized lowercase https prefix

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -449,3 +449,29 @@ func NewHTMLEscapeWriter(w io.Writer) io.Writer {
+	// url must not contain any prefix, so add https prefix
+	log.Infof("Prepending https:// to url: %s", url)
+	pURL = httpsText + url
+	return pURL
```
same, just 1 line instead of 2:
```
return httpsPrefix + url
```

though again I thought the linters would catch/not like that and force it to be HTTPSPrefix ?

olimpias committed:
JoinHostAndPort is renamed to NormalizeHostPort and updated. Some test cases are added.

ldemailly commented on fnet/network.go:
```diff
@@ -161,19 +161,12 @@ func ProxyToDestination(listenPort string, destination string) *net.TCPAddr {
+// when the original port binding input didn't specify an address
+func NormalizeHostPort(inputPort string, addr *net.TCPAddr) string {
+	urlHostPort := addr.String()
+	if strings.HasPrefix(inputPort, ":") || !strings.Contains(urlHostPort, ":") {
```
both should be inputPort; basically unless the input had a addr:port it will be changed to localhost:port

ldemailly replied:
see
https://github.com/istio/fortio/pull/257/files#diff-61a4e729b116ffe8a0303004555c636c

olimpias replied:
I see. I think, I got the idea :)

ldemailly commented on fnet/network_test.go:
```diff
@@ -169,6 +169,40 @@ func TestResolveIpV6(t *testing.T) {
+		{"8080", &net.TCPAddr{
+			IP:   []byte{192, 168, 2, 3},
+			Port: 8081,
+		}, "192.168.2.3:8081"},
```
this would be localhost unless you put 192.168.2.3:8081 as input (the tests need to include using addr:port)

olimpias committed:
More tests are added. Small change on NormalizeHostPort

ldemailly committed:
Merge branch 'master' into http_utils_test

ldemailly commented:
linters work now:

fhttp/http_utils.go:458:1:warning: comment on exported function AddHTTPS should be of the form "AddHTTPS ..." (golint)

ldemailly committed:
fixing lint issue

olimpias commented:
Thanks for your comments @ldemailly :)

ldemailly commented:
it's in 1.0 !
